### ISSUE TITLE
axios requests, when using http adapter, can't handle response body as object

### ISSUE THREAD

**Issue by ScottyMJacobson** [mswjs/node-request-interceptor#58]
I'd been using axios just fine to basically build a much simpler `nock`-type interface using `node-request-interceptor` (fantastic lib btw! thanks for providing it), but then i realized that in my _real_ tests, which will be consuming the interface i'm making, i'm using 

1) inside jest
2) using `axios.defaults.adapter = require('axios/lib/adapters/http');`.

so once i remembered and i set that adapter line (2^), my code, which was working beautifully, started throwing:

`TypeError [ERR_INVALID_ARG_TYPE]: The first argument must be one of type string, Buffer, ArrayBuffer, Array, or Array-like Object. Received type object`

Here's a repro file, which throws unless you assign a string to `body:` or if you comment out the `axios.defaults.adapter` line.

```
'use strict';

const { RequestInterceptor } = require('node-request-interceptor');
const withDefaultInterceptors = require('node-request-interceptor/lib/presets/default').default;

const interceptor = new RequestInterceptor(withDefaultInterceptors);

function handleResponse(request) {
    return {
        status: 200,
        headers: {
            some: 'header'
        },
        body: {
            valueFromObject: 'thisWorkedWithTheOtherAxiosAdapter'
        }
    };
}

interceptor.use(handleResponse);

const axios = require('axios');
axios.defaults.adapter = require('axios/lib/adapters/http');

describe('Repo my test that is breaking', () => {
    it('Returns the expecting string error', async () => {
        const iWontGetHere = await axios('http://theurlshouldntmatter.com');

        console.log(iWontGetHere);
    });
});

```

Thank you in advance! I tried `JSON.stringify()`ing the object, but then it escapes the JSON characters and `iWontGetHere.data` becomes a string rather than the object i would be expecting

**Comment by ScottyMJacobson** [comment]
ah! interesting! - slight update -

when i said 

>I tried JSON.stringify()ing the object, but then it escapes the JSON characters and iWontGetHere.data becomes a string rather than the object i would be expecting

It turns out `iWontGetHere.data` actually is indeed an object, and my _expected object that i'm using to match_, since it's passed by reference into my actual handler function, has been serialized into a string and so the matcher breaks

so effectively my actual use case is more like

(but there's more setup / i store the configs in an object and then do my own route matching, etc)

```
let retObj;

setRetObj(_retObj) {
  retObj = _retObj;
}

handler() {
  retObj.body = JSON.stringify(retObj.body)
  
  return retObj;
}
```

```
// test
const retObj = { body: { someKey: 'someVal' } };
setRetObj(retObj)
const r = await doRequest();
expect(r.data).toEqual(retObj.body)
```

so! i have a workaround! (or the way it's supposed to work in the first place? since `body` is documented as `body?: string`. 
🙂). i just `deepCopy` the object before JSON.stringifying the body.

maybe for consistency, it should also break when using the other adapter (i think it would be `axios` / `jsdom` / `xhrhttprequest`)

sorry for misreading my error message! (the one comparing the string to the object)

**Comment by kettanaito** [comment]
Hey, @ScottyMJacobson 👋 Thanks for reaching out with this.

I'm excited to see you building your own interception library on top of NRI! Looking forward to hear more feedback from you in the future.

Regarding this issue, I've tried your code above and the test passes for me without any additional setup.

```js
const { RequestInterceptor } = require('node-request-interceptor')
const withDefaultInterceptors = require('node-request-interceptor/lib/presets/default')
  .default

const interceptor = new RequestInterceptor(withDefaultInterceptors)

interceptor.use(() => {
  return {
    status: 200,
    headers: {
      some: 'header',
    },
    body: {
      valueFromObject: 'thisWorkedWithOtherAxiosAdapter',
    },
  }
})

const axios = require('axios')
axios.default.adapter = require('axios/lib/adapters/http')

test('example test', async () => {
  const response = await axios('http://mswjs.io')

  expect(response.status).toBe(200)
  expect(response.headers).toEqual({ some: 'header' })
  expect(response.data).toEqual({
    valueFromObject: 'thisWorkedWithOtherAxiosAdapter',
  })
})
```

```json
// package.json
"dependencies": {
  "axios": "^0.20.0",
  "jest": "^26.4.2",
  "node-request-interceptor": "^0.5.1"
}
```

Could you please update `node-request-interceptor` to the latest version just in case?

**Comment by ScottyMJacobson** [comment]
Hey @kettanaito - thanks for getting back to me!

In your repro file

`axios.default.adapter`

should actually be

`axios.defaults.adapter`

([relevant docs](https://github.com/axios/axios#global-axios-defaults))

Thanks!

**Comment by ScottyMJacobson** [comment]
and to reiterate from my second comment (i know my writing can be hard to follow sometimes)

according to your docs, it actually _shouldn't_ work in either case. so i would just like consistency (although i do like the convenience of being able to pass an object rather than being responsible for `JSON.stringify()` myself)

**Comment by kettanaito** [comment]
Hey, @ScottyMJacobson. I've looked into this issue once more and I believe I found the solution.

## What axios adapter affects

Without any adapters the `axios()` call constructs a request using `XMLHttpRequest` class (more precisely its polyfill in jsdom from Jest). However, when you enable the `axios/lib/adapters/http` adapter the requests are constructed using the `http.ClientRequest` class (regular requests in NodeJS). 

## The root cause

`XMLHttpRequest` and `http.ClientRequest` seem to be handling a response body differently. XHR can digest an object as the body value, while `ClientRequest` **always expects a string, or a Buffer-like object**. This means that when a `ClientRequest` is constructed with the object body you have in the code snippet you'd get the following error:

```
TypeError [ERR_INVALID_ARG_TYPE]: The first argument must be one of type string, Buffer, ArrayBuffer, Array, or Array-like Object. Received type object
```

> The error originates from `Buffer.from(mockecResponse.body)`, which is a step to convert a mocked response body into a Buffer to respond with.

## The solution

Stringify the mocked response body.

```js
interceptor.use(() => {
  return {
    status: 200,
    headers: {
      some: 'header',
    },
    body: JSON.stringify({ // notice this line
      valueFromObject: 'thisWorkedWithOtherAxiosAdapter',
    })
  }
})
```

This makes the test pass and the mocked response to be valid. This also behaves the same with and without the axios http adapter. 

Let me know if this helps.
